File: antoanlaodongfe/src/types/examTemplate.ts
import type { ExamType, TrainingGroup, QuestionType, DifficultyLevel, ApprovalStatus } from './enums';

export interface QuestionDistribution {
  topic_tag?: string;
  question_type?: QuestionType;
  difficulty?: DifficultyLevel;
  count: number;
}

export interface ExamTemplateRequest {
  name: string;
  exam_type: ExamType;
  training_group: TrainingGroup;
  occupation: string;
  skill_level: number;
  total_questions: number;
  duration_minutes: number;
  passing_score: number;
  distributions: QuestionDistribution[];
  excellent_threshold: number;
  good_threshold: number;
  average_threshold: number;
  created_by: string;
}

export interface ExamTemplateResponse extends ExamTemplateRequest {
  id: string;
  status: ApprovalStatus;
  created_at: string;
  updated_at: string;
}

export interface ExamTemplateListFilters {
  exam_type?: ExamType;
  status?: ApprovalStatus;
  occupation?: string;
  skill_level?: number;
  search?: string;
  page?: number;
  page_size?: number;
}

File: AnToanLaoDongBE/app/models/exam_template.py
from beanie import Document
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime, timezone

from app.models.enums import (
    ExamType, TrainingGroup, DifficultyLevel, QuestionType, ApprovalStatus,
)


class QuestionDistribution(BaseModel):
    topic_tag: Optional[str] = None
    question_type: Optional[QuestionType] = None
    difficulty: Optional[DifficultyLevel] = None
    count: int


class ExamTemplate(Document):
    name: str
    exam_type: ExamType
    training_group: TrainingGroup
    occupation: Optional[str] = None
    skill_level: Optional[int] = None

    total_questions: int
    duration_minutes: int
    passing_score: float

    distributions: list[QuestionDistribution] = []

    # Classification thresholds (on scale of 10)
    excellent_threshold: float = 9.0
    good_threshold: float = 7.0
    average_threshold: float = 5.0

    status: ApprovalStatus = ApprovalStatus.DRAFT
    requested_at: Optional[datetime] = None
    requested_to: Optional[str] = None
    requested_department_id: Optional[str] = None
    review_notes: Optional[str] = None
    created_by: str
    reviewed_by: Optional[str] = None
    approved_at: Optional[datetime] = None

    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    class Settings:
        name = "exam_templates"
        indexes = ["exam_type", "occupation", "skill_level", "status"]
